core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1581,6 +1650,178 @@ private void validateTableModelAndCreateTypeCast(
+
+        @SuppressWarnings("unchecked")
+        ObjHashSet<CharSequence> tableNames = (ObjHashSet<CharSequence>) cachedObjSet;
+        try {
+            tableNames.clear();
+            while (true) {
+                CharSequence tok = SqlUtil.fetchNext(lexer);
+                if (null == tok) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" expected a table name");
+                }
+                CharSequence tableName = GenericLexer.unhack(GenericLexer.unquote(tok));
+                if (tableName == null) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" expected a valid table name");
+                }
+                int status = engine.getStatus(executionContext.getCairoSecurityContext(), path, tableName, 0, tableName.length());
+                if (status != TableUtils.TABLE_EXISTS) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" '").put(tableName).put("' is not  a valid table");
+                }
+                tableNames.add(tableName);
+
+                tok = SqlUtil.fetchNext(lexer);
+                if (null == tok || Chars.equals(tok, ';')) {
+                    break;
+                }
+                if (!Chars.equals(tok, ',')) {
+                    throw SqlException.position(lexer.lastTokenPosition()).put(" expected ','");
+                }
+            }
+
+            for (int n = 0; n < tableNames.size(); n++) {
+                backupTable(tableNames.get(n), executionContext);
+            }
+
+            return compiledQuery.ofBackupTable();
+        } finally {
+            tableNames.clear();
+        }
+    }
+
+    private void setupBackupRenamePath() {
+        TimestampFormat format = configuration.getBackupDirTimestampFormat();
+        long epochMicros = configuration.getMicrosecondClock().getTicks();
+        int n = 0;
+        // TODO: There is a race here, two threads could try and create the same renamePath, only one will succeed the other will throw
+        // a CairoException. Maybe it should be serialised
+        do {
+            renamePath.of(configuration.getBackupRoot()).put(Files.SEPARATOR);
+            format.format(epochMicros, TimestampLocaleFactory.INSTANCE.getDefaultTimestampLocale(), null, renamePath);
+            if (n > 0) {
+                renamePath.put('.').put(n);
+            }
+            renamePath.put(Files.SEPARATOR).$();
+            n++;
+        } while (ff.exists(renamePath));
+        if (ff.mkdirs(renamePath, configuration.getMkDirMode()) != 0) {
+            throw CairoException.instance(ff.errno()).put("could not create [dir=").put(renamePath).put(']');
+        }
+    }
+
+    private transient String cachedTmpBackupRoot;
+
+    private void backupTable(@NotNull CharSequence tableName, @NotNull SqlExecutionContext executionContext) {
+        LOG.info().$("Starting backup of ").$(tableName).$();
+        if (null == cachedTmpBackupRoot) {
```

Comment:
Caching backup root like this can cause a potential side effect that server will not be able to back a table after backup directory was successfully created by server and then manually removed by user.

Checking dir for existence is a low overhead generally compared to creating a directory anyway.

Oh, looking down the code, `cloneMetaData` will create these directories anyway. May be we should not be creating directory upfront like this?